### ISSUE TITLE
Cannot resolve symbol 'styled'

### ISSUE THREAD

**Issue by jvanderen1** [ben-rogerson/twin.macro#533]
When using `twin.macro`, IntelliJ is telling me that `styled` is not a part of the package:

<img width="557" alt="Screen Shot 2021-09-10 at 1 08 43 PM" src="https://user-images.githubusercontent.com/22926257/132911730-c5792b2a-5ab9-4a70-b22e-4c8f2b64f030.png">

Are there any plugins I need or any additional configurations I should make?

**Comment by ben-rogerson** [comment]
Not sure about this one, I don't use that ide so perhaps someone who has used this before can help.

Just to confirm for testing, which jetbrains product are you using?

![image](https://user-images.githubusercontent.com/21288568/132926424-ef03a255-5c60-45d7-a281-79ba9cf211a2.png)

**Comment by jvanderen1** [comment]
I specifically am using IntelliJ. I can try to look into it a little bit more this weekend 👍🏻.

**Comment by jvanderen1** [comment]
I think I found the root cause. I'm looking at the `index.d.ts`, and I don't see any named export named `styled` from here:

```ts
import { ComponentType } from 'react'

export interface TwStyle {
  [key: string]: string | number | TwStyle
}

export const GlobalStyles: string

export type TemplateFn<R> = (
  strings: Readonly<TemplateStringsArray>,
  ...values: readonly string[]
) => R

export type TwFn = TemplateFn<TwStyle>

export type ThemeSearchFn<R> = (...values: readonly string[]) => R
export type ThemeSearchTaggedFn<R> = (
  strings: Readonly<TemplateStringsArray>
) => R

export type ThemeFn = <T = string>(arg?: string | TemplateStringsArray) => T

export type ScreenFn = <T = string>(
  screenValue: string | TemplateStringsArray
) => (styles?: string | TemplateStringsArray | TwStyle) => T

export type TwComponent<K extends keyof JSX.IntrinsicElements> = (
  props: JSX.IntrinsicElements[K]
) => JSX.Element

export type TwComponentMap = {
  [K in keyof JSX.IntrinsicElements]: TemplateFn<TwComponent<K>>
}

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
type TwComponentWrapper = <T extends ComponentType<any>>(
  component: T
) => TemplateFn<T>

declare const tw: TwFn & TwComponentMap & TwComponentWrapper
export default tw

declare module 'react' {
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  interface DOMAttributes<T> {
    tw?: string
  }
  interface HTMLAttributes<T> extends AriaAttributes, DOMAttributes<T> {
    cs?: string
  }
}

declare global {
  namespace JSX {
    interface IntrinsicAttributes {
      tw?: string
    }
  }
}

declare const theme: ThemeFn
declare const screen: ScreenFn

declare const globalStyles: Record<string, unknown>

export { theme, screen, globalStyles }
```

**Comment by jvanderen1** [comment]
In an ideal state, we should be using typescript for our code. Then, could have `microbundle` do the heavy lifting instead by looking at the compiled types automatically with a configuration like:

```json
{
  "name": "twin.macro",
  "main": "macro.js",
  "types": "dist/index.d.ts",
  "source": "src/macro.js",
  "scripts": {
    "build": "npm run build:macro",
    "build:macro": "microbundle --target node",
  },
}
```

But this would require us to convert all files into Typescript and possibly change CI/CD configurations. I would be more than happy to do the prior as a PR.

**Comment by ben-rogerson** [comment]
If you're using typescript, you'll need to setup your styled and css import types yourself as they are directly passed through from your chosen css-in-js framework.
I didn't see a .tsx on your file above so assumed you weren't using typescript here so I didn't suggest this originally.
If you have any ideas on how to automate this, I'm all ears :)
Another possibility is creating community type packages, eg: `@types/twin-styled-components` / `@types/emotion` etc.

Here's some examples of the typescript setups mentioned in the twin example readmes: [emotion](https://github.com/ben-rogerson/twin.examples/tree/master/vite-emotion-typescript#add-typescript-types) / [styled-components](https://github.com/ben-rogerson/twin.examples/tree/master/vite-styled-components-typescript#add-typescript-types)

**Comment by jvanderen1** [comment]
Yeah, JetBrains IDE's automatically look for types when trying to resolve exports 👍🏻 .

Microbundle has a way to generate typings based on the source file you give it. It automatically creates different bundle types based on the `source` inside your `package.json`, with the ability to configure the output via `main`, `typings`, etc. properties within the same `package.json`. Here is [microbundle's README](https://github.com/developit/microbundle#-installation--setup--) and checkout out this example generating typescript types: https://github.com/jvanderen1/typescript-microbundle-example

Maybe we could make twin.macro more robust by converting this project to Typescript and having the typescript compiler generate our types, including all supported flavors of CSS-in-JS? Again, I would love to contribute and give it a shot on my spare time.

**Comment by saeedeyvazy** [comment]
you must do this three steps

**Install react types**
`npm install -D @types/react`

**Install types for styled-components**
 add file named  twin.d.ts in root of src directory

i```
mport 'twin.macro'
import styledImport, { CSSProp, css as cssImport } from 'styled-components'

declare module 'twin.macro' {
  // The styled and css imports
  const styled: typeof styledImport
  const css: typeof cssImport
}

declare module 'react' {
  // The css prop
  interface HTMLAttributes<T> extends DOMAttributes<T> {
    css?: CSSProp
  }
  // The inline svg css prop
  interface SVGProps<T> extends SVGProps<SVGSVGElement> {
    css?: CSSProp
  }
}

// The 'as' prop on styled components
declare global {
  namespace JSX {
    interface IntrinsicAttributes<T> extends DOMAttributes<T> {
      as?: string | Element
    }
  }
}
```

Then add the following in your typescript config:

```
// tsconfig.json
{
  "files": ["twin.d.ts"],
  // or "include": ["twin.d.ts"],
}
```
